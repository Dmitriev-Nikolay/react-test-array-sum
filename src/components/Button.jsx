import React from 'react';

const Button = React.memo(({ setArray, arrayResult, allSum }) => {

    const updateArray = () => {
        setArray([...arrayResult, allSum]);
    };

    return (
        <>
            <button type="button" onClick={ updateArray }>Выполнить</button>
        </>
    );
});

export default Button;
