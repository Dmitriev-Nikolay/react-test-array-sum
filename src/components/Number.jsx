import React from 'react';

const Number = React.memo(({ name, change, valueInput }) => {

    return (
        <>
            <input
                className="inputs-block__item"
                type="number"
                min="1"
                name={ name }
                value={ valueInput }
                onChange={ change }
            />
        </>
    );
});

export default Number;