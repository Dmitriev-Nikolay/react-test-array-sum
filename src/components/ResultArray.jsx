import React from 'react';

const ResultArray = React.memo(({ result }) => {

    return (
        <section className="array-block">
            {
                result.length === 0 ?
                    <p className="array-block__start">Начинай</p>
                    :
                    <div className="array-block__result">
                        { 
                            result.map((num, i) => {
                                return (
                                    <p key={ `${ num }_${ i }` }>{ num }</p>
                                )
                            })
                        }
                    </div>
            }
        </section>
    );
});

export default ResultArray;
