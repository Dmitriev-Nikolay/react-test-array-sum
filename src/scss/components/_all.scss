@import './array';
@import './inputs';
@import './button';