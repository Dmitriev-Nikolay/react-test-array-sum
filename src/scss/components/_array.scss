@import "/src/scss/variables";
@import "/src/scss/libs/mixins";

.array-block {
    display: flex;
    justify-content: center;
    margin-bottom: 50px;
    max-width: 750px;

    &__start {
        padding: 25px 10px;
        font-size: 32px;
        border-radius: $borderRadius;
        border: $border;
    }

    &__result {
        display: flex;
        flex-wrap: wrap;
        border-radius: 10px;
        min-width: 150px;
        padding: 25px 10px;
        border: $border;
        font-size: 32px;

        p + p {
            margin-left: 8px;
        }

        p:after {
            content: ",";
        }
    }
}
