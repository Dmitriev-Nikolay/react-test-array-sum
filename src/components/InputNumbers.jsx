import React from 'react';

import { Number } from '../components';

const InputNumbers = React.memo(({ inputValueObj, handleChange }) => {

    return (
        <section className="inputs-block">
            <div className="inputs-block__items">
                <Number 
                    change={ handleChange } 
                    valueInput={ inputValueObj.firstValueInput } 
                    name='firstValueInput'
                />
                <Number 
                    change={ handleChange } 
                    valueInput={ inputValueObj.secondValueInput } 
                    name='secondValueInput'
                />
            </div>
        </section>
    );
});

export default InputNumbers;