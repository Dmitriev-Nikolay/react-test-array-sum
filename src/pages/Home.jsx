import React from 'react';

import { ResultArray, InputNumbers, Button } from '../components';

const Home = () => {
    const [resultOfNumbers, setIsResultOfNumbers] = React.useState([]);
    const [sumOfTwo, setIsSumOfTwo] = React.useState(0);
    const [inputValueObj, setInputValues] = React.useState({ firstValueInput: 1, secondValueInput: 1 });

    const handleChange = (event) => {
        setInputValues({ ...inputValueObj, [event.target.name]: +event.target.value })
    };

    const getArray = (arr) => {
        setIsResultOfNumbers(arr);
    };

    const getSum = React.useCallback((obj) => {
        let result = Object.values(obj).reduce((res, el) => res += el, 0);
        setIsSumOfTwo(result);
    }, []);

    React.useEffect(() => {
        getSum(inputValueObj);
    }, [getSum, inputValueObj]);

    return (
        <div className="content">
            <div className="container">
                <ResultArray
                    result={ resultOfNumbers }
                />
                <InputNumbers
                    handleChange={ handleChange }
                    inputValueObj={ inputValueObj }
                />
                <Button
                    setArray={ getArray }
                    arrayResult={ resultOfNumbers }
                    allSum={ sumOfTwo }
                />
            </div>
        </div>
    );
};

export default Home;