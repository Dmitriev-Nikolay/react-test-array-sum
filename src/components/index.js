import Number from './Number';
import InputNumbers from './InputNumbers';
import ResultArray from './ResultArray';
import Button from './Button';

export {
    Number,
    InputNumbers,
    ResultArray,
    Button,
};