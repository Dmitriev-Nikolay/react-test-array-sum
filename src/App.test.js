import { shallow, mount } from 'enzyme';
import React from 'react';

import { Home } from './pages';
import { ResultArray, InputNumbers, Number, Button } from './components';


//------------------------------ NUMBER ------------------------------//

const propsNumber = {
  name: 'firstValueInput',
  valueInput: '1',
};

const setUpNumber = (props) => shallow(<Number {...props} />);

describe('Number component', () => {

  /**
* Проверка на рендер одного инпута в компоненте
*/

  describe('Initial render', () => {
    const component = setUpNumber(propsNumber);
    it('Should render number element', () => {
      const number = component.find('input');
      expect(number).toHaveLength(1);
    });
  });

  /**
 * Проверка на дефолтное значение (undefined) пропса name
 */

  describe('name defaultProps', () => {
    it('Should use default name', () => {
      const result = Number.name;
      expect(result).toBe(undefined);
    });
  });

  /**
 * Проверка на дефолтное значение (undefined) пропса valueInput
 */

  describe('valueInput defaultProps', () => {
    it('Should use default valueInput', () => {
      const result = Number.valueInput;
      expect(result).toBe(undefined);
    });
  });

  /**
* Проверка обработчика события изменения поля ввода
*/

  describe('Number component onChange func', () => {
    it("check the onChange callback", () => {
      const onChange = jest.fn(),
        props = {
          value: "20",
          onChange
        },
        NumberComponent = mount(<Number {...props} />).find("input");
      NumberComponent.simulate("change", {
        target: { value: onChange(props.value) }
      });
      expect(onChange).toHaveBeenCalledWith(props.value);
    });
  });
});

//------------------------------ BUTTON ------------------------------//

const propsButton = {
  allSum: 0,
  arrayResult: [],
};

const setUpButton = (props) => mount(<Button {...props} />);

describe('Button component', () => {

  /**
* Проверка на рендер одной кнопки в компоненте
*/

  describe('Initial render', () => {
    const componentButton = setUpButton(propsButton);
    it('Should render button element', () => {
      const button = componentButton.find('button');
      expect(button).toHaveLength(1);
    });
  });

  /**
 * Проверка на дефолтное значение (undefined) пропса allsum
 */

  describe('allSum defaultProps', () => {
    it('Should use default allSum', () => {
      const props = {
        allsum: 0,
      },
        componentButton = mount(<Button {...props} />);
      expect(componentButton.prop("allsum")).toEqual(0);
    });
  });

  /**
 * Проверка на дефолтное значение (undefined) пропса valueInput
 */

  describe('arrayResult defaultProps', () => {
    it('Should use default arrayResult', () => {
      const props = {
        arrayResult: []
      },
        componentButton = mount(<Button {...props} />);
      expect(componentButton.prop("arrayResult")).toEqual([]);
    });
  });
});

//------------------------------ INPUTNUMBERS ------------------------------//

const propsInputNumbers = {
  inputValueObj: {
    firstValueInput: 1,
    secondValueInput: 1
  }
};

const setUpInputNumbers = (props) => shallow(<InputNumbers {...props} />);

describe('InputNumbers component', () => {

  /**
* Проверка на рендер 2-х компонентов Number
*/

  describe('Initial render', () => {
    const component = setUpInputNumbers(propsInputNumbers);
    it('Should render 2 elements of numbers', () => {
      const numbers = component.find(Number);
      expect(numbers).toHaveLength(2);
    });
  });
});

//------------------------------ RESULTARRAY ------------------------------//

const propsResultArray = {
  result: []
};

const setUpResultArray = (props) => shallow(<ResultArray {...props} />);

describe('ResultArray component', () => {

  /**
* Проверка на рендер заглушки с текстом, когда массив результатов пустой
*/

  describe('Initial render', () => {
    const component = setUpResultArray(propsResultArray);
    it('Should render stub with text', () => {
      const text = component.find('p');
      expect(text).toHaveLength(1);
    });
  });

  /**
* Проверка на рендер массива результатов, если его длина === 0
*/

  describe('Initial render', () => {
    const component = setUpResultArray(propsResultArray);
    it('Should render stub with text', () => {
      const text = component.find('div');
      expect(text).toHaveLength(0);
    });
  });
});

//------------------------------ HOME ------------------------------//

describe('Home component', () => {

  /**
* Проверка на рендер всего контента
*/

  describe('Initial render', () => {
    const HomeComponent = mount(<Home />);
    it('Should render all content', () => {
      const homeContent = HomeComponent.find(".content");
      expect(homeContent).toHaveLength(1);
    });
  });
});