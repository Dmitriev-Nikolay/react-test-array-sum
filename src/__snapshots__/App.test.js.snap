// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button component Should render Button component with props 1`] = `ShallowWrapper {}`;

exports[`Button component Should render Button component without props 1`] = `ShallowWrapper {}`;

exports[`Number component Should render Number component with props 1`] = `ShallowWrapper {}`;

exports[`Number component Should render Number component without props 1`] = `ShallowWrapper {}`;
